File: frontend/src/lib/api/collections.ts
import type { Collection, CollectionItem } from '$lib/types';

const BASE = '/api/collections';

export async function getCollections(categoryId?: number): Promise<Collection[]> {
	const url = categoryId != null ? `${BASE}/?category_id=${categoryId}` : `${BASE}/`;
	const res = await fetch(url);
	return res.json();
}

export async function getCollection(id: number): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}`);
	return res.json();
}

export async function createCollection(data: { title: string; category_id: number; description?: string }): Promise<Collection> {
	const res = await fetch(`${BASE}/`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function updateCollection(id: number, data: Partial<Collection>): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function deleteCollection(id: number): Promise<void> {
	await fetch(`${BASE}/${id}`, { method: 'DELETE' });
}

export async function activateCollection(id: number): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}/activate`, { method: 'POST' });
	return res.json();
}

export async function deactivateCollection(id: number): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}/deactivate`, { method: 'POST' });
	return res.json();
}

export async function archiveCollection(id: number): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}/archive`, { method: 'POST' });
	return res.json();
}

export async function unarchiveCollection(id: number): Promise<Collection> {
	const res = await fetch(`${BASE}/${id}/unarchive`, { method: 'POST' });
	return res.json();
}

export async function addItem(collectionId: number, data: { member_entity_type: string; member_entity_id: number; position?: number; status?: string; notes?: string; header?: string }): Promise<CollectionItem> {
	const res = await fetch(`${BASE}/${collectionId}/items`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || `Failed to add item (${res.status})`);
	}
	return res.json();
}

export async function updateItem(itemId: number, data: { position?: number; status?: string; notes?: string; header?: string | null }): Promise<CollectionItem> {
	const res = await fetch(`${BASE}/items/${itemId}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function removeItem(itemId: number): Promise<void> {
	await fetch(`${BASE}/items/${itemId}`, { method: 'DELETE' });
}

File: src/wellbegun/routers/collections.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from wellbegun.database import get_db
from wellbegun.schemas.collection import (
    CollectionCreate,
    CollectionItemCreate,
    CollectionItemOut,
    CollectionItemUpdate,
    CollectionOut,
    CollectionUpdate,
)
from wellbegun.services import collection_service

router = APIRouter(prefix="/collections", tags=["collections"])


@router.get("/", response_model=list[CollectionOut])
def list_collections(category_id: int | None = None, db: Session = Depends(get_db)):
    return collection_service.get_all(db, category_id=category_id)


@router.get("/{collection_id}", response_model=CollectionOut)
def get_collection(collection_id: int, db: Session = Depends(get_db)):
    collection = collection_service.get_by_id(db, collection_id)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


@router.post("/", response_model=CollectionOut, status_code=201)
def create_collection(data: CollectionCreate, db: Session = Depends(get_db)):
    return collection_service.create(
        db,
        title=data.title,
        category_id=data.category_id,
        description=data.description,
    )


@router.put("/{collection_id}", response_model=CollectionOut)
def update_collection(
    collection_id: int, data: CollectionUpdate, db: Session = Depends(get_db)
):
    updates = data.model_dump(exclude_unset=True)
    collection = collection_service.update(db, collection_id, **updates)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


@router.delete("/{collection_id}")
def delete_collection(collection_id: int, db: Session = Depends(get_db)):
    if not collection_service.delete(db, collection_id):
        raise HTTPException(status_code=404, detail="Collection not found")
    return {"ok": True}


# --- Lifecycle endpoints ---

@router.post("/{collection_id}/activate", response_model=CollectionOut)
def activate_collection(collection_id: int, db: Session = Depends(get_db)):
    collection = collection_service.activate(db, collection_id)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


@router.post("/{collection_id}/deactivate", response_model=CollectionOut)
def deactivate_collection(collection_id: int, db: Session = Depends(get_db)):
    collection = collection_service.deactivate(db, collection_id)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


@router.post("/{collection_id}/archive", response_model=CollectionOut)
def archive_collection(collection_id: int, db: Session = Depends(get_db)):
    collection = collection_service.archive(db, collection_id)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


@router.post("/{collection_id}/unarchive", response_model=CollectionOut)
def unarchive_collection(collection_id: int, db: Session = Depends(get_db)):
    collection = collection_service.unarchive(db, collection_id)
    if not collection:
        raise HTTPException(status_code=404, detail="Collection not found")
    return collection


# --- Item endpoints ---

@router.post("/{collection_id}/items", response_model=CollectionItemOut, status_code=201)
def add_item(
    collection_id: int, data: CollectionItemCreate, db: Session = Depends(get_db)
):
    try:
        return collection_service.add_item(
            db,
            collection_id=collection_id,
            member_entity_type=data.member_entity_type,
            member_entity_id=data.member_entity_id,
            position=data.position,
            status=data.status,
            notes=data.notes,
            header=data.header,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))


@router.put("/items/{item_id}", response_model=CollectionItemOut)
def update_item(item_id: int, data: CollectionItemUpdate, db: Session = Depends(get_db)):
    updates = data.model_dump(exclude_unset=True)
    try:
        item = collection_service.update_item(db, item_id, **updates)
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.delete("/items/{item_id}")
def remove_item(item_id: int, db: Session = Depends(get_db)):
    if not collection_service.remove_item(db, item_id):
        raise HTTPException(status_code=404, detail="Item not found")
    return {"ok": True}
